Signature Appearance can't be write to last page

Hi,

I  write a UnitTest that putting the signature appearance into the last page of PDF, but get a wrong result as below:
```
=== RUN   TestVisualSignLastPage
    sign_test.go:108: pdf total pages: 14
    sign_test.go:142: failed to create visual signature: page number 14 not found
--- FAIL: TestVisualSignLastPage (0.14s)
```
Test code as below:

``` golang
func TestVisualSignLastPage(t *testing.T) {
	_signer := &PdfSigner{}
	_signer.loadPKCS12("./cert.p12", "password")
	cert, sk, err := _signer.generateCert(UserCertInfo{
		Org:        []string{"Test Org"},
		Country:    []string{"CN"},
		CommonName: "TestCommonName",
		NotBefore:  time.Now(),
		NotAfter:   time.Now().AddDate(1, 0, 0),
	})
	if err != nil {
		t.Fail()
	}
	input_file, err := os.Open("./testfile16.pdf")
	if err != nil {
		t.Fail()
	}
	defer input_file.Close()

	output_file, err := os.Create("./output.pdf")
	if err != nil {
		t.Fail()
	}
	defer output_file.Close()

	finfo, err := input_file.Stat()
	if err != nil {
		t.Fail()
	}
	size := finfo.Size()

	rdr, err := pdf.NewReader(input_file, size)
	if err != nil {
		t.Fail()
	}
	lastPage := rdr.NumPage()
        t.Logf("pdf total pages: %d", lastPage)
	err = sign.Sign(input_file, output_file, rdr, size, sign.SignData{
		Signature: sign.SignDataSignature{
			Info: sign.SignDataSignatureInfo{
				Name:        "John Doe",
				Location:    "Somewhere on the globe",
				Reason:      "My season for siging this document",
				ContactInfo: "How you like",
				Date:        time.Now().Local(),
			},
			CertType:   sign.ApprovalSignature,
			DocMDPPerm: sign.AllowFillingExistingFormFieldsAndSignaturesPerms,
		},
		Signer:          sk,            // crypto.Signer
		DigestAlgorithm: crypto.SHA256, // hash algorithm for the digest creation
		Certificate:     cert,          // x509.Certificate
		Appearance: sign.Appearance{
			Visible:     true,
			LowerLeftX:  400,
			LowerLeftY:  50,
			UpperRightX: 600,
			UpperRightY: 125,
			Page:        uint32(lastPage),
			// Image:       signatureImage, // JPG or PNG image bytes
			// ImageAsWatermark: true,   // Optional: set to true to draw text over the image
		},
		// The follow options are likely to change in a future release
		//
		// cache revocation data when bulk signing
		RevocationData: revocation.InfoArchival{},
		// custom revocation lookup
		RevocationFunction: sign.DefaultEmbedRevocationStatusFunction,
	})
	if err != nil {
		t.Fatal(err)
	} else {
		t.Logf("Signed PDF written to output.pdf")
	}
}
```

--- a/sign/pdfvisualsignature.go
+++ b/sign/pdfvisualsignature.go
@@ -169,21 +169,35 @@ func (context *SignContext) createIncPageUpdate(pageNumber, annot uint32) ([]byt
 	return page_buffer.Bytes(), nil
 }
 
-// Helper function to find a page by its number.
+// Helper function to find a page by its number
 func findPageByNumber(pages pdf.Value, pageNumber uint32) (pdf.Value, error) {
+	page, remaining, err := findPageByNumberRec(pages, pageNumber)
+	if err != nil {
+		return pdf.Value{}, err
+	}
+	if remaining != 0 {
+		return pdf.Value{}, fmt.Errorf("page number %d not found", pageNumber)
+	}
+	return page, nil
+}
+
+// Internal recursive helper that returns the found page and the remaining page number to find.
+func findPageByNumberRec(pages pdf.Value, pageNumber uint32) (pdf.Value, uint32, error) {
 	if pages.Key("Type").Name() == "Pages" {
 		kids := pages.Key("Kids")
 		for i := 0; i < kids.Len(); i++ {
-			page, err := findPageByNumber(kids.Index(i), pageNumber)
-			if err == nil {
-				return page, nil
+			page, remaining, err := findPageByNumberRec(kids.Index(i), pageNumber)
+			if err == nil && remaining == 0 {
+				return page, 0, nil
 			}
+			pageNumber = remaining
 		}
+		return pdf.Value{}, pageNumber, fmt.Errorf("page number %d not found", pageNumber)
 	} else if pages.Key("Type").Name() == "Page" {
 		if pageNumber == 1 {
-			return pages, nil
+			return pages, 0, nil
 		}
-		pageNumber--
+		return pdf.Value{}, pageNumber - 1, nil
 	}
-	return pdf.Value{}, fmt.Errorf("page number %d not found", pageNumber)
+	return pdf.Value{}, pageNumber, fmt.Errorf("page number %d not found", pageNumber)
 }

--- a/sign/sign_test.go
+++ b/sign/sign_test.go
@@ -617,7 +617,7 @@ func TestSignPDFWithTwoImages(t *testing.T) {
 	verifySignedFile(t, secondSignature, filepath.Base(tbsFile))
 }
 
-// TestSignPDFWithWatermarkImage tests signing a PDF with a JPG image and text above
+// TestSignPDFWithWatermarkImageJPG tests signing a PDF with a JPG image and text above
 func TestSignPDFWithWatermarkImageJPG(t *testing.T) {
 	cert, pkey := loadCertificateAndKey(t)
 	inputFilePath := "../testfiles/testfile12.pdf"
@@ -749,7 +749,7 @@ func TestVisualSignLastPage(t *testing.T) {
 	if err != nil {
 		t.Fail()
 	}
-	lastPage := rdr.NumPage() - 1
+	lastPage := rdr.NumPage()
 	t.Logf("pdf total pages: %d", lastPage)
 	err = Sign(input_file, tmpfile, rdr, size, SignData{
 		Signature: SignDataSignature{
@@ -781,10 +781,5 @@ func TestVisualSignLastPage(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	err2 := os.Rename(tmpfile.Name(), "../testfiles/failed/"+originalFileName)
-	if err2 != nil {
-		t.Error(err2)
-	}
-
-	verifySignedFile(t, tmpfile, filepath.Base("../testfiles/testfile16.pdf"))
+	verifySignedFile(t, tmpfile, originalFileName)
 }

--- a/sign/sign_test.go
+++ b/sign/sign_test.go
@@ -720,3 +720,71 @@ func TestSignPDFWithWatermarkImagePNG(t *testing.T) {
 
 	verifySignedFile(t, tmpfile, originalFileName)
 }
+
+func TestVisualSignLastPage(t *testing.T) {
+	cert, pkey := loadCertificateAndKey(t)
+	inputFilePath := "../testfiles/testfile16.pdf"
+	input_file, err := os.Open(inputFilePath)
+	originalFileName := filepath.Base(inputFilePath)
+	if err != nil {
+		t.Fail()
+	}
+	defer input_file.Close()
+
+	tmpfile, err := os.CreateTemp("", t.Name())
+	if err != nil {
+		t.Fail()
+	}
+	if !testing.Verbose() {
+		defer os.Remove(tmpfile.Name())
+	}
+
+	finfo, err := input_file.Stat()
+	if err != nil {
+		t.Fail()
+	}
+	size := finfo.Size()
+
+	rdr, err := pdf.NewReader(input_file, size)
+	if err != nil {
+		t.Fail()
+	}
+	lastPage := rdr.NumPage() - 1
+	t.Logf("pdf total pages: %d", lastPage)
+	err = Sign(input_file, tmpfile, rdr, size, SignData{
+		Signature: SignDataSignature{
+			Info: SignDataSignatureInfo{
+				Name:        "John Doe",
+				Location:    "Somewhere on the globe",
+				Reason:      "My season for signing this document",
+				ContactInfo: "How you like",
+				Date:        time.Now().Local(),
+			},
+			CertType:   ApprovalSignature,
+			DocMDPPerm: AllowFillingExistingFormFieldsAndSignaturesPerms,
+		},
+		Signer:          pkey,          // crypto.Signer
+		DigestAlgorithm: crypto.SHA256, // hash algorithm for the digest creation
+		Certificate:     cert,          // x509.Certificate
+		Appearance: Appearance{
+			Visible:     true,
+			LowerLeftX:  400,
+			LowerLeftY:  50,
+			UpperRightX: 600,
+			UpperRightY: 125,
+			Page:        uint32(lastPage),
+		},
+		RevocationData:     revocation.InfoArchival{},
+		RevocationFunction: DefaultEmbedRevocationStatusFunction,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err2 := os.Rename(tmpfile.Name(), "../testfiles/failed/"+originalFileName)
+	if err2 != nil {
+		t.Error(err2)
+	}
+
+	verifySignedFile(t, tmpfile, filepath.Base("../testfiles/testfile16.pdf"))
+}

